web/interceptor/tokens_test.go

```diff
@@ -0,0 +1,159 @@
+package interceptor_test
+
+import (
+	"context"
+	"net"
+	"testing"
+	"time"
+
+	"github.com/golang-jwt/jwt"
+	"github.com/quickfeed/quickfeed/ci"
+	"github.com/quickfeed/quickfeed/internal/qtest"
+	"github.com/quickfeed/quickfeed/qf"
+	"github.com/quickfeed/quickfeed/qlog"
+	"github.com/quickfeed/quickfeed/scm"
+	"github.com/quickfeed/quickfeed/web"
+	"github.com/quickfeed/quickfeed/web/auth"
+	"github.com/quickfeed/quickfeed/web/interceptor"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/test/bufconn"
+)
+
+func TestRefreshTokens(t *testing.T) {
+	db, cleanup := qtest.TestDB(t)
+	defer cleanup()
+	logger := qlog.Logger(t)
+	ags := web.NewQuickFeedService(logger.Desugar(), db, scm.TestSCMManager(), web.BaseHookOptions{}, &ci.Local{})
+
+	tm, err := auth.NewTokenManager(db, "test")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	lis := bufconn.Listen(BufSize)
+	bufDialer := func(context.Context, string) (net.Conn, error) {
+		return lis.Dial()
+	}
+	opt := grpc.ChainUnaryInterceptor(
+		interceptor.UnaryUserVerifier(logger, tm),
+		interceptor.TokenRefresher(logger, tm),
+	)
+	s := grpc.NewServer(opt) // skipcq: GO-S0902
+	qf.RegisterQuickFeedServiceServer(s, ags)
+
+	go func() {
+		if err := s.Serve(lis); err != nil {
+			t.Errorf("Server exited with error: %v", err)
+			return
+		}
+	}()
+
+	admin := qtest.CreateFakeUser(t, db, 1)
+	user := qtest.CreateFakeUser(t, db, 56)
+
+	ctx := context.Background()
+	conn, err := grpc.DialContext(ctx, "bufnet", grpc.WithContextDialer(bufDialer), grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		t.Fatalf("Failed to dial bufnet: %v", err)
+	}
+	defer conn.Close()
+
+	client := qf.NewQuickFeedServiceClient(conn)
+
+	adminCookie, err := tm.NewAuthCookie(admin.ID)
+	if err != nil {
+		t.Fatal(err)
+	}
+	userCookie, err := tm.NewAuthCookie(user.ID)
+	if err != nil {
+		t.Fatal(err)
+	}
+	adminCtx := qtest.WithAuthCookie(ctx, interceptor.AuthTokenString(adminCookie.Value))
+	userCtx := qtest.WithAuthCookie(ctx, interceptor.AuthTokenString(userCookie.Value))
+	adminClaims := &auth.Claims{
+		UserID: admin.ID,
+		Admin:  true,
+		StandardClaims: jwt.StandardClaims{
+			ExpiresAt: time.Now().Add(1 * time.Minute).Unix(),
+		},
+	}
+	userClaims := &auth.Claims{
+		UserID: user.ID,
+		StandardClaims: jwt.StandardClaims{
+			ExpiresAt: time.Now().Add(1 * time.Minute).Unix(),
+		},
+	}
+	if tm.UpdateRequired(adminClaims) || tm.UpdateRequired(userClaims) {
+		t.Error("No users should be in the token update list at the start")
+	}
+	if _, err := client.GetUsers(adminCtx, &qf.Void{}); err != nil {
+		t.Fatal(err)
+	}
+	if tm.UpdateRequired(userClaims) {
```

Comment:
Shouldn't this be `adminClaims`? Or is this just testing that the `userClaims` wasn't impacted by the above `GetUsers(adminCtx ...` thing?

Reply:
Looks like it is reasonable to check both users, updated.